sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorReceiver.java

```diff
@@ -15,77 +17,145 @@
+public class ReactorReceiver implements AmqpReceiveLink, AsyncAutoCloseable {
     private final String entityPath;
     private final Receiver receiver;
     private final ReceiveLinkHandler handler;
     private final TokenManager tokenManager;
     private final ReactorDispatcher dispatcher;
     private final Disposable subscriptions;
     private final AtomicBoolean isDisposed = new AtomicBoolean();
+    private final Sinks.Empty<Void> isClosedMono = Sinks.empty();
     private final Flux<Message> messagesProcessor;
+    private final AmqpRetryOptions retryOptions;
     private final ClientLogger logger = new ClientLogger(ReactorReceiver.class);
     private final Flux<AmqpEndpointState> endpointStates;
 
     private final AtomicReference<Supplier<Integer>> creditSupplier = new AtomicReference<>();
 
-    protected ReactorReceiver(String entityPath, Receiver receiver, ReceiveLinkHandler handler,
-        TokenManager tokenManager, ReactorDispatcher dispatcher) {
+    protected ReactorReceiver(AmqpConnection amqpConnection, String entityPath, Receiver receiver,
+        ReceiveLinkHandler handler, TokenManager tokenManager, ReactorDispatcher dispatcher,
+        AmqpRetryOptions retryOptions) {
         this.entityPath = entityPath;
         this.receiver = receiver;
         this.handler = handler;
         this.tokenManager = tokenManager;
         this.dispatcher = dispatcher;
+        // Delivered messages are not published on another scheduler because we want the settlement method that happens
+        // in decodeDelivery to take place and since proton-j is not thread safe, it could end up with hundreds of
+        // backed up deliveries waiting to be settled. (Which, consequently, ends up in a FAIL_OVERFLOW error from
+        // the handler.
         this.messagesProcessor = this.handler.getDeliveredMessages()
-            .map(this::decodeDelivery)
-            .doOnNext(next -> {
-                if (receiver.getRemoteCredit() == 0 && !isDisposed.get()) {
-                    final Supplier<Integer> supplier = creditSupplier.get();
-                    if (supplier == null) {
-                        return;
+            .flatMap(delivery -> {
+                return Mono.create(sink -> {
+                    try {
+                        this.dispatcher.invoke(() -> {
+                            final Message message = decodeDelivery(delivery);
+                            final int creditsLeft = receiver.getRemoteCredit();
+
+                            if (creditsLeft > 0) {
+                                sink.success(message);
+                                return;
+                            }
+
+                            final Supplier<Integer> supplier = creditSupplier.get();
+                            final Integer credits = supplier.get();
+
+                            if (credits != null && credits > 0) {
+                                logger.info("connectionId[{}] linkName[{}] adding credits[{}]",
```

Comment:
Would this kick in for every message, wondering is `.info` will fill the logs ?

Reply:
It kicks in far less than you'd imagine. We only check when there are no credits on the line. And from my experience, we request a pretty high number. It's one of the few indications that aren't verbose which shows you we're still pushing items on the link.